Addition of snyk-security workflow to enable the Snyk platform from a single GitHub Action

EricFernandezSnyk opened the pull request:
<!--
IMPORTANT:

This repository contains configuration for what users see when they click on the `Actions` tab and the setup page for Code Scanning.

It is not:
* A playground to try out scripts
* A place for you to create a workflow for your repository
-->

## Pre-requisites

- [x] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [x] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [ ] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [x] Should specify least privileged [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully.

**For _CI_ workflows, the workflow:**

- [ ] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [ ] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [ ] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [ ] Packaging workflows should run on `release` with `types: [ created ]`.
- [ ] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**For _Code Scanning_ workflows, the workflow:**

- [ ] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/code-scanning).
- [ ] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [ ] `name`: Name of the Code Scanning integration.
  - [ ] `creator`: Name of the organization/user producing the Code Scanning integration.
  - [ ] `description`: Short description of the Code Scanning integration.
  - [ ] `categories`: Array of languages supported by the Code Scanning integration.
  - [ ] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [ ] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- [ ] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [ ] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [ ] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [ ] Automation and CI workflows cannot be dependent on a paid service or product.

EricFernandezSnyk committed:
Create snyk-security.properties.json

EricFernandezSnyk committed:
Create snyk-security.yml

EricFernandezSnyk committed:
Merge branch 'actions:main' into main

EricFernandezSnyk committed:
Update snyk-security.yml

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,77 @@
+      - name: Set up Snyk CLI to check for security issues
+        # Snyk can be used to break the build when it detects security issues.
+        # In this case we want to upload the SAST issues to GitHub Code Scanning
+        uses: snyk/actions/setup@master
```
Can we have commit sha instead of a moving master branch?

EricFernandezSnyk replied:
Corrected

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,77 @@
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Snyk CLI to check for security issues
+        # Snyk can be used to break the build when it detects security issues.
```
This doc comment looks more helpful at the top

Phantsure replied:
If here it says that it should break the code then why at line 59 and 64 do you pass true for not failing?

EricFernandezSnyk replied:
Moved to top and made `|| true` optional

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,77 @@
+      - name: Upload result to GitHub Code Scanning
+        uses: github/codeql-action/upload-sarif@v2
+        with:
+          sarif_file: snyk-code.sarif
```
There is a possibility that this point would reach without creating sarif file. Can we fail at above step(line 59) if sarif file is not able to be created?

EricFernandezSnyk replied:
Corrected

EricFernandezSnyk committed:
Fix mispelling

Co-authored-by: Sampark Sharma <phantsure@github.com>

EricFernandezSnyk committed:
Apply comments from PR

- Moved documentation link to the top
- Made `|| true` optional
- Added commit SHA for the Snyk GitHub Action

EricFernandezSnyk reviewed (commented):
Updated Pull Request based on comments, thanks @Phantsure 😄

mattlovestech commented:
> Updated Pull Request based on comments, thanks @Phantsure 😄

Can you apply to the Technology Partnership program [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner) and also confirm that this github username is associated with Snyk by sending an email to partnerships@github.com from your snyk email?

EricFernandezSnyk commented:
> > Updated Pull Request based on comments, thanks @Phantsure 😄
> 
> Can you apply to the Technology Partnership program [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner) and also confirm that this github username is associated with Snyk by sending an email to [partnerships@github.com](mailto:partnerships@github.com) from your snyk email?

Thanks I have now done this 🙂

mattlovestech commented:
@Phantsure Snyk is now a partner and this username was confirmed.

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+          # This is where you will need to introduce the Snyk API token created with your Snyk account
+          SNYK_TOKEN: ${{ secrets.SNYK_TOKEN }}
+        
+        # Runs Snyk Code (SAST) analysis and uploads result into GitHub. 
+        # Use || true to not fail the pipeline      
+      - name: Snyk Code test 
```
```suggestion
        # Runs Snyk Code (SAST) analysis and uploads result into GitHub.
        # Use || true to not fail the pipeline
      - name: Snyk Code test
```

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+      - name: Snyk Open Source monitor
+        run: snyk monitor --all-projects
+        
+        # Runs Snyk Infrastructure as Code (IaC) analysis and uploads result to Snyk. 
```
```suggestion
        # Runs Snyk Infrastructure as Code (IaC) analysis and uploads result to Snyk.
```

Phantsure committed:
Merge branch 'main' into main

Phantsure commented:
I have pointed out some and other linting errors can be fixed with 
```
pip3 install pre-commit
pre-commit run --all-files --show-diff-on-failure --color always
```

EricFernandezSnyk committed:
Remove empty space

Co-authored-by: Sampark Sharma <phantsure@github.com>

EricFernandezSnyk committed:
Remove empty space in line end

Co-authored-by: Sampark Sharma <phantsure@github.com>

EricFernandezSnyk committed:
Update Categories

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+        # Snyk can be used to break the build when it detects security issues.
+        # In this case we want to upload the SAST issues to GitHub Code Scanning
+        uses: snyk/actions/setup@806182742461562b67788a64410098c9d9b96adb
+ 
```
```suggestion

```

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+        env:
+          # This is where you will need to introduce the Snyk API token created with your Snyk account
+          SNYK_TOKEN: ${{ secrets.SNYK_TOKEN }}
+        
```
```suggestion

```

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+        # Runs Snyk Open Source (SCA) analysis and uploads result to Snyk.
+      - name: Snyk Open Source monitor
+        run: snyk monitor --all-projects
+        
```
```suggestion

```

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+        # Use || true to not fail the pipeline
+      - name: Snyk Code test
+        run: snyk code test --sarif > snyk-code.sarif # || true
+        
```
```suggestion

```

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+        # Use || true to not fail the pipeline.
+      - name: Snyk IaC test and report
+        run: snyk iac test --report # || true
+        
```
```suggestion

```

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+        # Runs Snyk Container (Container and SCA) analysis and uploads result to Snyk.
+      - name: Snyk Container monitor
+        run: snyk container monitor your/image-to-test --file=Dockerfile
+        
```
```suggestion

```

Phantsure commented on code-scanning/snyk-security.yml:
```diff
@@ -0,0 +1,79 @@
+        #- uses: actions/setup-node@v3
+        #  with:
+        #    node-version: 16
+        
```
```suggestion

```

Phantsure reviewed (changes requested):
linting changes needed

EricFernandezSnyk committed:
Updated after running pre-commit linting

Phantsure commented:
@EricFernandezSnyk Good to merge?

Jcpayvoice23 commented:
I don't want no trouble for you if you don't mind me asking then who owns that platform because I remember doing a platform like that engineering backing 2014-15
